File: rust/src/utils/direction.rs
//! Direction and coordinate utilities
//!
//! Common types for working with 2D positions and directions.

/// A 2D point with signed coordinates
pub type Point = (i32, i32);

/// Cardinal directions
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum Direction {
    North,
    South,
    East,
    West,
}

impl Direction {
    /// Get all four cardinal directions
    pub fn all() -> [Direction; 4] {
        [
            Direction::North,
            Direction::South,
            Direction::East,
            Direction::West,
        ]
    }

    /// Get the (dx, dy) delta for this direction
    /// North is -y, South is +y, West is -x, East is +x
    pub fn delta(&self) -> (i32, i32) {
        match self {
            Direction::North => (0, -1),
            Direction::South => (0, 1),
            Direction::East => (1, 0),
            Direction::West => (-1, 0),
        }
    }

    /// Turn left (counter-clockwise)
    pub fn turn_left(&self) -> Self {
        match self {
            Direction::North => Direction::West,
            Direction::West => Direction::South,
            Direction::South => Direction::East,
            Direction::East => Direction::North,
        }
    }

    /// Turn right (clockwise)
    pub fn turn_right(&self) -> Self {
        match self {
            Direction::North => Direction::East,
            Direction::East => Direction::South,
            Direction::South => Direction::West,
            Direction::West => Direction::North,
        }
    }

    /// Turn around (180 degrees)
    pub fn turn_around(&self) -> Self {
        match self {
            Direction::North => Direction::South,
            Direction::South => Direction::North,
            Direction::East => Direction::West,
            Direction::West => Direction::East,
        }
    }

    /// Parse from a character (^, v, <, >, N, S, E, W)
    pub fn from_char(c: char) -> Option<Self> {
        match c {
            '^' | 'N' | 'n' | 'U' | 'u' => Some(Direction::North),
            'v' | 'V' | 'S' | 's' | 'D' | 'd' => Some(Direction::South),
            '>' | 'E' | 'e' | 'R' | 'r' => Some(Direction::East),
            '<' | 'W' | 'w' | 'L' | 'l' => Some(Direction::West),
            _ => None,
        }
    }

    /// Convert to a character representation
    pub fn to_char(&self) -> char {
        match self {
            Direction::North => '^',
            Direction::South => 'v',
            Direction::East => '>',
            Direction::West => '<',
        }
    }
}

/// Move a point in a direction
pub fn move_point(point: Point, direction: Direction) -> Point {
    let (dx, dy) = direction.delta();
    (point.0 + dx, point.1 + dy)
}

/// Move a point in a direction by a given distance
pub fn move_point_by(point: Point, direction: Direction, distance: i32) -> Point {
    let (dx, dy) = direction.delta();
    (point.0 + dx * distance, point.1 + dy * distance)
}

/// Calculate Manhattan distance between two points
pub fn manhattan_distance(a: Point, b: Point) -> i32 {
    (a.0 - b.0).abs() + (a.1 - b.1).abs()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_direction_turn() {
        assert_eq!(Direction::North.turn_right(), Direction::East);
        assert_eq!(Direction::North.turn_left(), Direction::West);
        assert_eq!(Direction::North.turn_around(), Direction::South);
    }

    #[test]
    fn test_manhattan_distance() {
        assert_eq!(manhattan_distance((0, 0), (3, 4)), 7);
        assert_eq!(manhattan_distance((1, 1), (1, 1)), 0);
    }
}

File: python/src/aoc2025/utils/direction.py
"""Direction and coordinate utilities.

Common types for working with 2D and 3D positions and directions.
"""

import math
from enum import Enum

# Type alias for a 2D point
Point = tuple[int, int]

# Type alias for a 3D point
Point3 = tuple[int, int, int]


class Direction(Enum):
    """Cardinal directions."""

    NORTH = (0, -1)
    SOUTH = (0, 1)
    EAST = (1, 0)
    WEST = (-1, 0)

    @classmethod
    def all(cls) -> list["Direction"]:
        """Get all four cardinal directions."""
        return [cls.NORTH, cls.SOUTH, cls.EAST, cls.WEST]

    @property
    def delta(self) -> tuple[int, int]:
        """Get the (dx, dy) delta for this direction."""
        return self.value

    def turn_left(self) -> "Direction":
        """Turn left (counter-clockwise)."""
        turns = {
            Direction.NORTH: Direction.WEST,
            Direction.WEST: Direction.SOUTH,
            Direction.SOUTH: Direction.EAST,
            Direction.EAST: Direction.NORTH,
        }
        return turns[self]

    def turn_right(self) -> "Direction":
        """Turn right (clockwise)."""
        turns = {
            Direction.NORTH: Direction.EAST,
            Direction.EAST: Direction.SOUTH,
            Direction.SOUTH: Direction.WEST,
            Direction.WEST: Direction.NORTH,
        }
        return turns[self]

    def turn_around(self) -> "Direction":
        """Turn around (180 degrees)."""
        turns = {
            Direction.NORTH: Direction.SOUTH,
            Direction.SOUTH: Direction.NORTH,
            Direction.EAST: Direction.WEST,
            Direction.WEST: Direction.EAST,
        }
        return turns[self]

    @classmethod
    def from_char(cls, c: str) -> "Direction | None":
        """Parse from a character (^, v, <, >, N, S, E, W)."""
        mapping = {
            "^": cls.NORTH,
            "N": cls.NORTH,
            "n": cls.NORTH,
            "U": cls.NORTH,
            "u": cls.NORTH,
            "v": cls.SOUTH,
            "V": cls.SOUTH,
            "S": cls.SOUTH,
            "s": cls.SOUTH,
            "D": cls.SOUTH,
            "d": cls.SOUTH,
            ">": cls.EAST,
            "E": cls.EAST,
            "e": cls.EAST,
            "R": cls.EAST,
            "r": cls.EAST,
            "<": cls.WEST,
            "W": cls.WEST,
            "w": cls.WEST,
            "L": cls.WEST,
            "l": cls.WEST,
        }
        return mapping.get(c)

    def to_char(self) -> str:
        """Convert to a character representation."""
        chars = {
            Direction.NORTH: "^",
            Direction.SOUTH: "v",
            Direction.EAST: ">",
            Direction.WEST: "<",
        }
        return chars[self]


def move_point(point: Point, direction: Direction, distance: int = 1) -> Point:
    """Move a point in a direction by a given distance."""
    dx, dy = direction.delta
    return (point[0] + dx * distance, point[1] + dy * distance)


def manhattan_distance(a: Point, b: Point) -> int:
    """Calculate Manhattan distance between two points."""
    return abs(a[0] - b[0]) + abs(a[1] - b[1])


def euclidean_difference(a: Point3, b: Point3) -> float:
    return math.sqrt((a[0] - b[0]) ** 2 + (a[1] - b[1]) ** 2 + (a[2] - b[2]) ** 2)
